README.md
# Spin Circle Bottom Bar

An easy to implement Spin Circle Bottom Navigation Bar for Flutter Applications.

## Current Features

* Initial Release for Spin Circle Bottom Bar
* Customizable Tab Icon and Text for both Active and Inactive State
* Customize colors for Spin Circle
* Customize Bottom Bar and Spin Circle Item Count

## Demo
![](https://github.com/retroportalstudio/spincircle_bottom_bar/blob/master/spincircle_bottom_bar.gif)

## Usage
To Use, simply add SpinCircleBottomBarHolder to your Scaffold's body, as follows:

```dart
class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Scaffold(
        appBar: AppBar(
          title: Text("RetroPortal Studio"),
          backgroundColor: Colors.deepOrangeAccent,
        ),
        //Adding SpinCircleBottomBarHolder to body of Scaffold
        body: SpinCircleBottomBarHolder(
          bottomNavigationBar: SCBottomBarDetails(
              circleColors: [Colors.white, Colors.orange, Colors.redAccent],
              iconTheme: IconThemeData(color: Colors.black45),
              activeIconTheme: IconThemeData(color: Colors.orange),
              backgroundColor: Colors.white,
              titleStyle: TextStyle(color: Colors.black45,fontSize: 12),
              activeTitleStyle: TextStyle(color: Colors.black,fontSize: 12,fontWeight: FontWeight.bold),
              actionButtonDetails: SCActionButtonDetails(
                  color: Colors.redAccent,
                  icon: Icon(
                    Icons.expand_less,
                    color: Colors.white,
                  ),
                  elevation: 2),
              elevation: 2.0,
              items: [
                // Suggested count : 4
                SCBottomBarItem(icon: Icons.verified_user, title: "User", onPressed: () {}),
                SCBottomBarItem(icon: Icons.supervised_user_circle, title: "Details", onPressed: () {}),
                SCBottomBarItem(icon: Icons.notifications, title: "Notifications", onPressed: () {}),
                SCBottomBarItem(icon: Icons.details, title: "New Data", onPressed: () {}),
              ],
              circleItems: [
                //Suggested Count: 3
                SCItem(icon: Icon(Icons.add), onPressed: () {}),
                SCItem(icon: Icon(Icons.print), onPressed: () {}),
                SCItem(icon: Icon(Icons.map), onPressed: () {}),
              ],
              bnbHeight: 80 // Suggested Height 80
          ),
          child: Container(
            color: Colors.orangeAccent.withAlpha(55),
            child: Center(child: Text("Isn't this Awesome!"),),
          ),
        ),
      ),
    );
  }
}

```

## Roadmap
Plans to add more customizations.

## License
[MIT](https://choosealicense.com/licenses/mit/)

CHANGELOG.md
## [1.0.0] - June 08, 2020

* Initial Release for Spin Circle Bottom Bar
* Customizable Tab Icon and Text for both Active and Inactive State
* Customize colors for Spin Circle
* Customize Bottom Bar and Spin Circle Item Count

test/spincirclebottombar_test.dart
//import 'package:flutter_test/flutter_test.dart';
//
//import 'package:spincirclebottombar/spincircle_bottom_bar.dart';
//
//void main() {
//  test('adds one to input values', () {
//    final calculator = Calculator();
//    expect(calculator.addOne(2), 3);
//    expect(calculator.addOne(-7), -6);
//    expect(calculator.addOne(0), 1);
//    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//  });
//}


lib/modals.dart
import 'package:flutter/material.dart';


class SCBottomBarDetails {
  Color backgroundColor;
  double elevation;
  List<Color> circleColors;
  IconThemeData activeIconTheme;
  IconThemeData iconTheme;
  TextStyle activeTitleStyle;
  TextStyle titleStyle;
  List<SCBottomBarItem> items;
  List<SCItem> circleItems;
  SCActionButtonDetails actionButtonDetails;
  double bnbHeight;

  SCBottomBarDetails(
      {@required this.items, @required this.circleItems,this.bnbHeight,this.actionButtonDetails, this.activeIconTheme, this.iconTheme, this.activeTitleStyle, this.titleStyle, this.circleColors,this.backgroundColor,this.elevation});
}

class SCActionButtonDetails{
  Color color;
  Icon icon;
  double elevation;

  SCActionButtonDetails({@required this.color,@required this.icon,@required this.elevation});
}

class SCItem {
  Icon icon;
  Function onPressed;
  SCItem({@required this.icon, @required this.onPressed});
}

class SCBottomBarItem {
  IconData activeIcon;
  IconData icon;
  String title;
  Function onPressed;

  SCBottomBarItem({this.activeIcon, @required this.icon, this.title, @required this.onPressed});
}


lib/spincircle_bottom_bar.dart
library spincircle_bottom_bar;

import 'package:flutter/material.dart';
import 'package:spincircle_bottom_bar/modals.dart';

class SpinCircleBottomBarHolder extends StatelessWidget {
  final SCBottomBarDetails bottomNavigationBar;
  final Widget child;

  const SpinCircleBottomBarHolder({Key key, @required this.bottomNavigationBar, @required this.child}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Stack(
      fit: StackFit.expand,
      children: <Widget>[
        Column(
          children: <Widget>[
            Expanded(child: child),
            Container(
              height: bottomNavigationBar.bnbHeight??80,
            )
          ],
        ),
        Positioned(
            bottom: 0,
            left: 0,
            child: SpinCircleBottomBar(
              bottomNavigationBar: bottomNavigationBar,
            ))
      ],
    );
  }
}

class SpinCircleBottomBar extends StatefulWidget {
  final SCBottomBarDetails bottomNavigationBar;

  const SpinCircleBottomBar({Key key, this.bottomNavigationBar}) : super(key: key);

  @override
  _SpinCircleBottomBarState createState() => _SpinCircleBottomBarState();
}

enum ExpansionStatus { OPEN, CLOSE, IDLE }

class _SpinCircleBottomBarState extends State<SpinCircleBottomBar> {
  ExpansionStatus expansionStatus = ExpansionStatus.IDLE;
  SCBottomBarDetails expandableBottomBarDetails;
  int activeIndex = 0;

  @override
  void initState() {
    super.initState();
    expandableBottomBarDetails = widget.bottomNavigationBar;
    expandableBottomBarDetails.items.insert((expandableBottomBarDetails.items.length / 2).floor(), null);
    expandableBottomBarDetails.circleColors = expandableBottomBarDetails.circleColors ?? [Colors.white, Colors.blue, Colors.red];
    expandableBottomBarDetails.actionButtonDetails = expandableBottomBarDetails.actionButtonDetails ??
        SCActionButtonDetails(
            color: Colors.blue,
            icon: Icon(
              Icons.add,
              color: Colors.white,
            ),
            elevation: 0);
  }

  @override
  Widget build(BuildContext context) {
    final width = MediaQuery.of(context).size.width;

    final double bottomBarHeight = expandableBottomBarDetails.bnbHeight ?? 80;
    final IconThemeData iconTheme = expandableBottomBarDetails.iconTheme ?? IconThemeData(color: Colors.black45);
    final IconThemeData activeIconTheme = expandableBottomBarDetails.activeIconTheme ?? IconThemeData(color: Colors.black);
    final TextStyle textStyle = expandableBottomBarDetails.titleStyle ?? TextStyle(color: Colors.black45, fontWeight: FontWeight.normal, fontSize: 12);
    final TextStyle activeTextStyle = expandableBottomBarDetails.activeTitleStyle ?? TextStyle(color: Colors.black, fontWeight: FontWeight.bold, fontSize: 12);
    final SCActionButtonDetails actionButtonDetails = expandableBottomBarDetails.actionButtonDetails;
    bool shouldOpen = expansionStatus == ExpansionStatus.OPEN;
    return Container(
        height: bottomBarHeight * 2,
        width: width,
        alignment: Alignment.center,
        child: Stack(
          fit: StackFit.expand,
          overflow: Overflow.visible,
          children: <Widget>[
            if (expansionStatus != ExpansionStatus.IDLE) ...[
              Container(
                  width: width,
                  height: bottomBarHeight * 2,
                  child: Stack(
                    children: <Widget>[
                      TweenAnimationBuilder(
                        tween: Tween<double>(begin: shouldOpen ? -3.14 : 0, end: shouldOpen ? 0 : -3.14),
                        curve: Curves.easeInOutQuad,
                        duration: Duration(milliseconds: shouldOpen ? 500 : 800),
                        builder: (BuildContext context, double value, Widget child) {
                          return Transform.rotate(
                            angle: value,
                            child: child,
                            alignment: Alignment.bottomCenter,
                          );
                        },
                        child: EmptyLayer(
                          radius: bottomBarHeight * 2,
                          color: expandableBottomBarDetails.circleColors[2],
                        ),
                      ),
                      TweenAnimationBuilder(
                        tween: Tween<double>(begin: shouldOpen ? -3.14 : 0, end: shouldOpen ? 0 : -3.14),
                        curve: Curves.easeInOutQuad,
                        duration: Duration(milliseconds: 600),
                        builder: (BuildContext context, double value, Widget child) {
                          return Transform.rotate(
                            angle: value,
                            child: child,
                            alignment: Alignment.bottomCenter,
                          );
                        },
                        child: EmptyLayer(
                          radius: bottomBarHeight * 2,
                          color: expandableBottomBarDetails.circleColors[1],
                        ),
                      ),
                      TweenAnimationBuilder(
                        tween: Tween<double>(begin: shouldOpen ? -3.14 : 0, end: shouldOpen ? 0 : -3.14),
                        duration: Duration(milliseconds: shouldOpen ? 800 : 500),
                        curve: Curves.easeInOutQuad,
                        builder: (BuildContext context, double value, Widget child) {
                          return Transform.rotate(
                            angle: value,
                            child: child,
                            alignment: Alignment.bottomCenter,
                          );
                        },
                        child: PrimaryCircle(
                          circleItems: expandableBottomBarDetails.circleItems,
                          radius: bottomBarHeight * 2,
                          color: expandableBottomBarDetails.circleColors[0],
                        ),
                      ),
                    ],
                  ))
            ],
            Positioned(
              bottom: 0,
              child: Container(
                height: bottomBarHeight,
                width: width,
                decoration: BoxDecoration(
                    color: expandableBottomBarDetails.backgroundColor ?? Colors.white,
                    boxShadow: [BoxShadow(color: Colors.black.withAlpha(55), offset: Offset(0, expandableBottomBarDetails.elevation ?? 0), blurRadius: 10)]),
                child: Center(
                  child: Padding(
                    padding: const EdgeInsets.all(8.0),
                    child: Row(
                      crossAxisAlignment: CrossAxisAlignment.center,
                      mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                      children: expandableBottomBarDetails.items.asMap().entries.map((entry) {
                        int index = entry.key;
                        SCBottomBarItem itemDetails = entry.value;
                        bool isActive = activeIndex == index;
                        return Flexible(
                            child: itemDetails != null
                                ? GestureDetector(
                                    onTap: () {
                                      setState(() {
                                        activeIndex = index;
                                      });
                                      itemDetails.onPressed();
                                    },
                                    child: Center(
                                      child: Column(
                                        mainAxisSize: MainAxisSize.max,
                                        mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                                        children: <Widget>[
                                          Icon(
                                            isActive ? itemDetails.activeIcon ?? itemDetails.icon : itemDetails.icon,
                                            color: isActive ? activeIconTheme.color : iconTheme.color,
                                            size: isActive ? activeIconTheme.size : iconTheme.size,
                                          ),
                                          itemDetails.title != null ? Text(itemDetails.title, style: isActive ? activeTextStyle : textStyle) : Center()
                                        ],
                                      ),
                                    ),
                                  )
                                : Center());
                      }).toList(),
                    ),
                  ),
                ),
              ),
            ),
            Container(
                height: bottomBarHeight * 2,
                width: MediaQuery.of(context).size.width,
                alignment: Alignment.center,
                child: TweenAnimationBuilder(
                  tween: Tween<double>(begin:shouldOpen?0.0:6.28319, end: shouldOpen?6.28319:0),
                  duration: Duration(milliseconds: 250),
                  builder: (context, angle, child) {
                    return Transform.rotate(
                        angle: expansionStatus== ExpansionStatus.IDLE?0.0:angle,
                        alignment: Alignment.center,
                        child: child);
                  },
                  child: FloatingActionButton(
                      elevation: actionButtonDetails.elevation,
                      backgroundColor: actionButtonDetails.color,
                      child: shouldOpen ? Icon(Icons.close) : actionButtonDetails.icon,
                      onPressed: () {
                        if (expansionStatus == ExpansionStatus.IDLE) {
                          this.setState(() {
                            expansionStatus = ExpansionStatus.OPEN;
                          });
                        } else {
                          this.setState(() {
                            expansionStatus = (expansionStatus == ExpansionStatus.OPEN) ? ExpansionStatus.CLOSE : ExpansionStatus.OPEN;
                          });
                        }
                      }),
                )),
          ],
        ));
  }
}

class PrimaryCircle extends StatelessWidget {
  final List<SCItem> circleItems;
  final Color color;
  final double radius;

  const PrimaryCircle({Key key, @required this.color, @required this.radius, @required this.circleItems}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    double radianGap = 3.14159 / circleItems.length;
    double start = radianGap / 2;
    return ClipRect(
      child: Align(
        heightFactor: 0.5,
        alignment: Alignment.topCenter,
        child: Container(
          width: radius,
          height: radius,
          decoration: BoxDecoration(
            shape: BoxShape.circle,
            color: color,
          ),
          child: Center(
            child: Stack(
              fit: StackFit.expand,
              overflow: Overflow.visible,
              children: circleItems.asMap().entries.map((entry) {
                SCItem value = entry.value;
                return Transform.translate(
                    offset: Offset.fromDirection(-(start + (entry.key * radianGap)), radius / 3),
                    child: GestureDetector(
                      onTap: value.onPressed,
                      child: value.icon,
                    ));
              }).toList(),
            ),
          ),
        ),
      ),
    );
  }
}

class EmptyLayer extends StatelessWidget {
  final Color color;
  final double radius;

  const EmptyLayer({Key key, this.color, this.radius}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ClipRect(
      child: Align(
        heightFactor: 0.5,
        alignment: Alignment.topCenter,
        child: Container(
          width: radius,
          height: radius,
          decoration: BoxDecoration(
            shape: BoxShape.circle,
            color: color,
          ),
          child: Center(),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:spincircle_bottom_bar/modals.dart';
import 'package:spincircle_bottom_bar/spincircle_bottom_bar.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Spin Circle Bottom Bar',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SafeArea(
      child: Scaffold(
        appBar: AppBar(
          title: Text("RetroPortal Studio"),
          backgroundColor: Colors.deepOrangeAccent,
        ),
        //Adding SpinCircleBottomBarHolder to body of Scaffold
        body: SpinCircleBottomBarHolder(
          bottomNavigationBar: SCBottomBarDetails(
              circleColors: [Colors.white, Colors.orange, Colors.redAccent],
              iconTheme: IconThemeData(color: Colors.black45),
              activeIconTheme: IconThemeData(color: Colors.orange),
              backgroundColor: Colors.white,
              titleStyle: TextStyle(color: Colors.black45,fontSize: 12),
              activeTitleStyle: TextStyle(color: Colors.black,fontSize: 12,fontWeight: FontWeight.bold),
              actionButtonDetails: SCActionButtonDetails(
                  color: Colors.redAccent,
                  icon: Icon(
                    Icons.expand_less,
                    color: Colors.white,
                  ),
                  elevation: 2),
              elevation: 2.0,
              items: [
                // Suggested count : 4
                SCBottomBarItem(icon: Icons.verified_user, title: "User", onPressed: () {}),
                SCBottomBarItem(icon: Icons.supervised_user_circle, title: "Details", onPressed: () {}),
                SCBottomBarItem(icon: Icons.notifications, title: "Notifications", onPressed: () {}),
                SCBottomBarItem(icon: Icons.details, title: "New Data", onPressed: () {}),
              ],
              circleItems: [
                //Suggested Count: 3
                SCItem(icon: Icon(Icons.add), onPressed: () {}),
                SCItem(icon: Icon(Icons.print), onPressed: () {}),
                SCItem(icon: Icon(Icons.map), onPressed: () {}),
              ],
              bnbHeight: 80 // Suggested Height 80
          ),
          child: Container(
            color: Colors.orangeAccent.withAlpha(55),
            child: Center(child: Text("Isn't this Awesome!"),),
          ),
        ),
      ),
    );
  }
}

